Sources/HLSLocalizerUtils.swift

```diff
@@ -0,0 +1,369 @@
+        self.replaceOccurrences(of: segmentUrl, with: relativeLocalPath, options: [], range: NSMakeRange(0, self.length))
+    }
+}
+
+extension DTGSelectionOptions: CustomStringConvertible {
+    public var description: String {
+        return """
+        Video: height=\(videoHeight ?? -1) width=\(videoWidth ?? -1) codecs=\(videoCodecs ?? []) bitrates=\(videoBitrates)
+        Audio: all=\(allAudioLanguages) list=\(audioLanguages ?? []) codecs=\(audioCodecs ?? [])
+        Text: all=\(allTextLanguages) list=\(textLanguages ?? [])
+        """
+    }
+    
+    func fullVideoCodecPriority() -> [CodecTag] {
+        return DTGSelectionOptions.fullCodecPriority(requestedTags: (videoCodecs ?? []).map {$0.tag}, allowedTags: allowedVideoCodecTags)
+    }
+    
+    func fullAudioCodecPriority() -> [CodecTag] {
+        return DTGSelectionOptions.fullCodecPriority(requestedTags: (audioCodecs ?? []).map {$0.tag}, allowedTags: allowedAudioCodecTags)
+    }
+    
+    static func fullCodecPriority(requestedTags: [CodecTag], allowedTags: [CodecTag]) -> [CodecTag] {
+        var codecPriority = requestedTags
+        for codec in allowedTags {
+            if !codecPriority.contains(codec) {
+                codecPriority.append(codec)
+            }
+        }
+        return codecPriority
+    }
+}
+
+extension TrackCodec: CustomStringConvertible {
+    public var description: String {
+        return tag
+    }
+    
+    static let audioCodecs: [TrackCodec] = {
+        return defaultAudioCodecOrder
+    }()
+    
+    static let videoCodecs: [TrackCodec] = {
+        return defaultVideoCodecOrder
+    }()
+    
+    static let defaultVideoCodecOrder: [TrackCodec] = [.hevc, .avc1]
+    
+    static let defaultAudioCodecOrder: [TrackCodec] = [.eac3, .ac3, .mp4a]
+    
+    var tag: CodecTag {
+        switch self {
+        case .avc1: return "avc1"
+        case .hevc: return "hvc1"
+        case .mp4a: return "mp4a"
+        case .ac3: return "ac-3"
+        case .eac3: return "ec-3"
+        }
+    }
+    
+    
+    func isAllowed(with options: DTGSelectionOptions) -> Bool {
+        switch self {
+        case .avc1: return true
+        case .hevc: return CodecSupport.hardwareHEVC || CodecSupport.softwareHEVC && options.allowInefficientCodecs
+        case .mp4a: return true
+        case .ac3: return CodecSupport.ac3
+        case .eac3: return CodecSupport.ec3
+        }
+    }
+}
+
+// Util
+func filter(streams: [M3U8Stream], 
+                        sortOrder: (M3U8Stream, M3U8Stream) -> Bool?, 
+                        filter: (M3U8Stream) -> Bool) -> [M3U8Stream] {
+    
+    if streams.count < 2 {
+        return streams
+    }
+    
+    let sorted = streams.stableSorted(by: sortOrder)
+    
+    let filtered = sorted.filter( filter )
+    
+    if filtered.isEmpty {
+        if let s = sorted.last {
+            return [s]
+        } else {
+            return []
```

Comment:
The else can be removed, will never get there.

Reply:
Fixed